server/bastionai_app/src/main.rs

```diff
@@ -17,25 +21,36 @@ mod remote_torch {
 }
 use remote_torch::remote_torch_server::{RemoteTorch, RemoteTorchServer};
 use remote_torch::{
-    Chunk, Devices, Empty, Metric, Optimizers, Reference, References, TestConfig, TrainConfig,
+    Chunk, ClientInfo, Devices, Empty, Metric, Optimizers, Reference, References, TestConfig,
+    TrainConfig,
 };
 
+mod telemetry;
+use telemetry::TelemetryEventProps;
+
 mod storage;
 use storage::{Artifact, Dataset, Module};
 
 mod utils;
 use utils::*;
 
+use crate::storage::SizedObjectsBytes;
+
 struct BastionAIServer {
     modules: RwLock<HashMap<Uuid, Artifact<Module>>>,
+    modules_hashes: RwLock<HashMap<Uuid, Arc<(String, Option<ClientInfo>)>>>,
     datasets: RwLock<HashMap<Uuid, Artifact<Dataset>>>,
+    datasets_hashes: RwLock<HashMap<Uuid, Arc<(String, Option<ClientInfo>)>>>,
+
 }
 
 impl BastionAIServer {
     pub fn new() -> Self {
         BastionAIServer {
             modules: RwLock::new(HashMap::new()),
+            modules_hashes: RwLock::new(HashMap::new()),
```

Comment:
Not sure why we need to store them but if this is really needed, maybe we could use the hashes as identifiers instead of the uuids?

Reply:
We could.
It will go away once we switch to using hashes.

Reply:
Sounds like a plan

Reply:
Yeah, for now we keep it, in order not to complicate the code.
Ping me once we decide to change it to hashes as ids and then I'll work on it.